File: CPP/algorithms/arrays/remove_duplicates.cpp
/*
Find duplicates in O(n) time and O(1) extra space
*/
/*
Input: n=7 , array[]={1, 2, 3, 6, 3, 6, 1}
Output: 1, 3, 6
*/
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

void duplicate(vector<int> &arr)
{
    map<int, int> duplicates;
    for (int i = 0; i < arr.size(); i++)
    {
        duplicates[arr[i]]++;
    }
    for (const auto &it : duplicates)
    {
        if (it.second > 1)
        {
            cout << it.first << " ";
        }
    }
}

void print_vector(const vector<int> &vec)
{
    for (int num : vec)
    {
        cout << num << " ";
    }
    cout << std::endl;
}

int main()
{
    vector<int> arr = {2, 1, 5, 4, 1, 3, 5};
    vector<int> arr2 = {2, 3, 8, 3, 13};

    cout << "arr: ";
    print_vector(arr);
    cout << "arr2: ";
    print_vector(arr2);

    duplicate(arr);
    return 0;
}

File: Python/algorithms/arrays/remove_duplicates.py
"""
    Finds duplicates in O(n) time and O(1) extra space.
    
    Note: This method modifies the input list and assumes numbers are > 0.
    """

from collections import Counter
from typing import List

def find_duplicates(arr: List[int]) -> List[int]:

    freq_map = Counter(arr)

    duplicates = [item for item, count in freq_map.items() if count > 1]

    return duplicates

if __name__ == "__main__":

    arr1 = [1, 2, 3, 6, 3, 6, 1]

    arr2 = [2, 1, 5, 4, 1, 3, 5]

    arr3 = [2, 3, 8, 3, 13]
    
    print(f"Original list: {arr1}")

    print(f"Duplicates: {find_duplicates(arr1)}")

    print("-" * 20)
    

    
    print(f"Original list: {arr2}")

    print(f"Duplicates: {find_duplicates(arr2)}")

    print("-" * 20)



    print(f"Original list: {arr3}")

    print(f"Duplicates: {find_duplicates(arr3)}")

    print("-" * 20)